CMakeLists.txt

```diff
@@ -30,6 +30,23 @@ project (sigc++)
 
 set( CMAKE_CXX_STANDARD 17 )
 
+# Turn on warnings for MSVC
+if (MSVC)
+    # Remove the CMake default of /W3 because when you add /W4, MSVC will complain
+    # about two warning level flags
+    string(REGEX REPLACE "(^|[ \t])/W[0-9]($|[ \t])" "\\1\\2" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
+    string(REGEX REPLACE "(^|[ \t])/W[0-9]($|[ \t])" "\\1\\2" CMAKE_CXX_FLAGS
+    "${CMAKE_CXX_FLAGS}")
+elseif(CXX_COMPILER_ID)
+endif()
+
+# Add compiler warning flags & turn warnings into errors
+add_compile_options(
```

Comment:
Wow. This really does seem to be the official way to do it. That syntax is hard to like.

I quite like that, in the autotools build, using mm-common, it figures out which of the command line options are actually acceptable to the compiler. Anyway, I'm just grumbling - this seems to be the correct cmake way.

Reply:
CMake has something similar to that for *some* command-line flags (e.g. [setting the C++ standard you want to use](https://cmake.org/cmake/help/v3.1/prop_tgt/CXX_STANDARD.html)), but not warnings...

The other syntax that could be used is something like this:

```CMake
if (MSVC)
   add_compile_options(/W4 /W3)
elseif(GNU)
   add_compile_options(-pedantic -Wall -Wextra ...)
elseif(Clang OR AppleClang)
   add_compile_options(-pedantic -Wall -Wextra ...)
endif()
```

That is probably clearer? As you say, those generator expressions aren't nice...